hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java

```diff
@@ -461,25 +470,136 @@ public void run() {
         }
 
         private void rebalanceLeaderships() {
-            Map<RaftEndpoint, CPMember> members = getMembers();
+            Collection<CPMember> allMembers = getMembers().values();
+            Map<CPMember, Entry<Integer, Collection<CPGroupId>>> leadershipPriorityMap
+                    = getLeadershipPriorityMap(allMembers);
+
             Collection<CPGroupId> groupIds = getCpGroupIds();
+            List<CPGroupSummary> cpGroupSummaries
+                    = groupIds.stream()
+                    .map(this::getCpGroup)
+                    .collect(Collectors.toList());
+
+            // Transferring leaderships from the lowest priority CP members
+            Set<CPMember> priorityCPMembers
+                    = rebalanceLeadershipsByPriority(leadershipPriorityMap, cpGroupSummaries);
+
+            // If there are several CP members with the same priority level, we should rebalance leadership evenly among them
+            if (priorityCPMembers.size() > 1) {
```

Comment:
I think this condition removed. For example:

- I have 5 members, 2 have 99 and 3 have 0 priority.
- I promote 2 new members with 99 priority.
- I would expect my groups to be rebalanced evenly.

Reply:
Maybe we can have a test case for this scenario.

Reply:
Thanks, @ramizdundar for your comments!
Unfortunately, the current implementation could only rebalance leadership within the existing CP group members.  With the existing API, there is no mechanism to replace one CP member with another one. The only way to change CP group members is to remove the CP member https://docs.hazelcast.com/hazelcast/5.1/cp-subsystem/management#removing-a-cp-member and promote a new one.

My suggestion is don’t change the CP group members' pre-selection logic before the release, and leave the decision to the next iteration/release.

To summarise the limitations of the current implementation:
- a CP group can only be formed from CP members with the lowest priority
- new CP members (promoted members) with the highest priority will not be counted for existing groups when rebalancing

Reply:
Thanks for the answer.